mem/buffers.go

```diff
@@ -0,0 +1,217 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/*
+Package mem does memory things
+*/
+package mem
+
+import (
+	"io"
+	"sync/atomic"
+)
+
+type BufferSlice []*Buffer
+
+type Buffer struct {
+	data  []byte
+	refs  *atomic.Int32
+	free  func([]byte)
+	freed bool
+}
+
+func NewBuffer(data []byte, free func([]byte)) *Buffer {
+	b := &Buffer{data: data, refs: new(atomic.Int32), free: free}
+	b.refs.Add(1)
+	return b
+}
+
+func Copy(data []byte, pool BufferPool) *Buffer {
+	buf := pool.Get(len(data))
+	copy(buf, data)
+	return NewBuffer(buf, pool.Put)
+}
+
+func (b *Buffer) ReadOnlyData() []byte {
+	if b.freed {
+		panic("Cannot read freed buffer")
+	}
+	return b.data
+}
+
+func (b *Buffer) Ref() *Buffer {
+	if b.freed {
+		panic("Cannot ref freed buffer")
+	}
+
+	b.refs.Add(1)
+	return &Buffer{
+		data: b.data,
+		refs: b.refs,
+		free: b.free,
+	}
+}
+
+func (b *Buffer) Free() {
+	if b.freed {
+		return
+	}
+
+	b.freed = true
+	refs := b.refs.Add(-1)
+	if refs == 0 && b.free != nil {
+		b.free(b.data)
+	}
+	b.data = nil
+}
+
+func (b *Buffer) Len() int {
+	// Convenience: io.Reader returns (n int, err error), and n is often checked
+	// before err is checked. To mimic this, Len() should work on nil Buffers.
+	if b == nil {
+		return 0
+	}
+	return len(b.ReadOnlyData())
+}
+
+func (b *Buffer) Split(n int) *Buffer {
+	if b.freed {
+		panic("Cannot split freed buffer")
+	}
+
+	b.refs.Add(1)
+
+	data := b.data
+	var free func([]byte)
+	if f := b.free; f != nil {
+		free = func(_ []byte) {
+			f(data)
+		}
+	}
+
+	b.data = data[:n]
+	b.free = nil
+
+	return &Buffer{
+		data: data[n:],
+		refs: b.refs,
+		free: free,
+	}
+}
+
+type Writer struct {
+	buffers *BufferSlice
+	pool    BufferPool
+}
+
+func (s *Writer) Write(p []byte) (n int, err error) {
+	buf := s.pool.Get(len(p))
+	n = copy(buf, p)
+
+	*s.buffers = append(*s.buffers, NewBuffer(buf, s.pool.Put))
+
+	return n, nil
+}
+
+func NewWriter(buffers *BufferSlice, pool BufferPool) *Writer {
```

Comment:
The point of passing a `BufferSlice` to the `Writer` is to eventually get to the written data, right?

Would a `Writer` method on a `BufferSlice` make sense, analogous to the `Reader` method, and the `Writer` method would take a `BufferPool` as argument.

Reply:
All the methods on `BufferSlice` are direct, instead of on `*BufferSlice`, e.g.:
```go
func (s BufferSlice) Materialize() []byte {
	out := make([]byte, s.Len())
	s.CopyTo(out)
	return out
}
```
In order for the `Writer` to be able to append to the `BufferSlice` it needs to be a pointer method. It would mean defining mixed pointer/non-pointer receiver methods, which is why I avoided it. The alternative would be to define _all_ `BufferSlice` methods on the pointer receiver, but it feels like that makes it less obvious that `BufferSlice` is meant to be immutable. I also do like the ergonomics of using `NewWriter` like this:
```go
var out mem.BufferSlice
w := mem.NewWriter(&out, pool)
```
vs this
```go
var out mem.BufferSlice
w := (&out).NewWriter(pool)
```
(^ the above example doesn't use `out := new(mem.BufferSlice)` because the codec API uses `mem.BufferSlice`, not `*mem.BufferSlice`)